Dataype generic implementation of {From,To}Row

fredefox opened the pull request:
This is my first time ever doing data-type generic programming, so
please review this carefully.  I conferred with [the documentation in
base](http://hackage.haskell.org/package/base-4.12.0.0/docs/GHC-Generics.html#g:10) and I also looked at the analoguous definition in the package
[`binary`](http://hackage.haskell.org/package/binary).  It was surprisingly easy to implement, but perhaps it
should be tested some more before merging in.  I've just quickly
tested it with the code-base I'm working on.

Note that the current implementation can only derive the generic
instance for product-types (not sum-types).

fredefox commented:
The `@since` tag on the default implementation incorrectly gets attached to the instance method. I don't know how (if possible) to fix this.

nurpax commented:
Code looks alright to me. Couple of questions:

- add a motivating example in the PR and docs
- does this have a perf impact? see https://nurpax.github.io/posts/2013-08-17-sqlite-simple-benchmarking.html for a benchmark

fredefox commented:
I suspect it should only have a negligible impact on compilation time for code that uses this feature. If at all.

fredefox commented:
The bench-mark you're referring to is this https://github.com/nurpax/db-bench/tree/master/haskell right?

nurpax commented:
Yep!

fredefox commented:
I've updated the haddock documentation and the changelog.

fredefox commented:
I'm attaching the benchmarking output I got from running your `bench` with sqlite-simple versions [0.4.16.0 ](http://hackage.haskell.org/package/sqlite-simple-0.4.16.0) and [0.4.16.2](https://github.com/fredefox/sqlite-simple/commit/66eae0fe008d1aee4357ed8dedb2080ee6514490) respectively:

[sqlite-simple-0-4-16-0.txt](https://github.com/nurpax/sqlite-simple/files/2461661/sqlite-simple-0-4-16-0.txt)
[sqlite-simple-0-4-16-2.txt](https://github.com/nurpax/sqlite-simple/files/2461662/sqlite-simple-0-4-16-2.txt)

If there's anything else. Please let me know.

fredefox commented:
I also made a change to your db-util where I added a test-case for the data-type shown below. The result are attached.

[wide-0-4-16.txt](https://github.com/nurpax/sqlite-simple/files/2462387/wide-0-4-16.txt)
[wide-0-4-17.txt](https://github.com/nurpax/sqlite-simple/files/2462388/wide-0-4-17.txt)

I haven't pushed the changes I made to the utility because I had to comment out the stuff related to the other sql back ends to get a-building.

```
data Wide = Wide
  [Char]
  Word
  Text
  (Maybe [Char])
  (Maybe Word)
  (Maybe Text)
  Integer
  Int
  Float
  Double
  Bool

deriving stock    instance Generic Wide
deriving anyclass instance NFData  Wide
#if MIN_VERSION_sqlite_simple(0,4,17)
deriving anyclass instance SQL.FromRow Wide
deriving anyclass instance SQL.ToRow   Wide
#else
instance SQL.FromRow Wide where
  fromRow = Wide <$> f <*> f <*> f <*> f <*> f <*> f <*> f <*> f <*> f <*> f <*> f
    where
    f :: SQL.FromField a => SQL.RowParser a
    f = SQL.field
instance SQL.ToRow Wide where
  toRow (Wide a0 a1 a2 a3 a4 a5 a6 a7 a8 a9 a10)
    = [f a0, f a1, f a2, f a3, f a4, f a5, f a6, f a7, f a8, f a9, f a10]
    where
    f :: SQL.ToField a => a -> SQL.SQLData
    f = SQL.toField
#endif
```

dmjio commented:
+1, would like to see this in as well.

heatherleaf commented:
+1, anything happening with this?

fredefox commented:
@heatherleaf you could have a look at the CI. I think that's what was bothering @nurpax. I'm not sure if the CI script is up-to-date tho...

nurpax commented:
Sorry folks, I just haven’t really worked on my Haskell libs in the past year or so. Too many interesting side projects and I’ve done a lot less Haskell recently. 

I do have a great candidate to maintain my Haskell libs, just need to hand him the permissions to work on these repos.

fredefox commented:
oh and @heatherleaf if you feel uneasy about depending on our own fork you can just copy the functionality into a separate module in your project. I believe I left a commit somewhere detailing how to. You could link it here for @dmjio and others...

heatherleaf commented:
> oh and @heatherleaf if you feel uneasy about depending on a separate fork you can just copy the functionality into a separate module in your project. I believe I left a commit somewhere detailing how to. You could link it here for @dmjio and others...

Thanks for the tip, I reverted to your old code now. 

@dmjio FYI, until this PR is sorted out, you can perhaps be inspired by this commit:
https://github.com/MUSTE-Project/MULLE/commit/6667fec5177f96f99ac318713597dbd7ac222002

The important module is `Database.SQLite.Simple.FromRow.Generic`, and using that you can write (see `Muste.Web.Database.Types`):

    instance ToRow User where
        toRow = genericToRow
    instance FromRow User where
        fromRow = genericFromRow

instead of:

    deriving anyclass instance ToRow   User
    deriving anyclass instance FromRow User

(which you would with this PR)

fredefox commented:
`deriving anyclass` just sounds so much better ;)

fredefox commented:
Critical public service announcement to anyone who've used snippets from this PR. Please see https://github.com/nurpax/sqlite-simple/commit/734ae4c618b448d1bd59d68142cc3f4b24e4bc8a with a fix for an inifinite loop arising from the use of a generically derived version of `ToRow`.

fredefox commented:
This should be good to merge now.

sigrlami commented:
Thanks, I'll take a look over the weekend, if everything ok, I'll merge it. On May 30, 2019 17:02, Frederik Hanghøj Iversen <notifications@github.com> wrote:This should be good to merge now.

—You are receiving this because you are subscribed to this thread.Reply to this email directly, view it on GitHub, or mute the thread.

evanrelf commented:
Anything holding this PR back at this point?

fredefox commented:
Hi @evanrelf. I force-pushed a change that fixed the merge conflict. As far as I know this can just be merged as-is. But I'm not really the maintainer, so it kinda depends on them I suppose. I think I've addressed all the concerns mentioned so far in the PR.

fredefox commented:
@evanrelf perhaps we can meet up and celebrate the 2 year birthday of this PR when the date rolls around :D

sigrlami commented:
I had time this weekend to test out. everything works. Hackage release will be in about 5 days, I need to work out additional tests and CI corrections with new functionality.
